R/subscript-tables.R

```diff
@@ -389,5 +401,118 @@ nameDimensionAttributes <- function(x)
         names(dimnames.x) <- dim.names
         dimnames(x) <- dimnames.x
     }
+    # attr(x, "dim.types") <- type.per.dim
     return(x)
 }
+
+# Return the number of dimensions for a given question type.
+# This can either be 1 or 2.
+#' @param question.types A vector of characters corresponding to question types
+#' @noRd
+questionDimension <- function(question.types) {
+    q.dims <- rep(1, length(question.types))
+    q.dims[question.types %in% c("PickOneMulti", "PickAnyGrid", "NumberGrid")] <- 2
+    q.dims
+}
+
+# Determine new questiontypes by comparing the new QTable to the original
+#' @param new The newly-subscripted QTable
+#' @param original The original QTable pre-subscripting
+#' @noRd
+getUpdatedQuestionTypes <- function(new, original) {
+    orig.q.types <- attr(original, "questiontypes")
+    if (is.null(orig.q.types))
+        return(NULL)
+
+    if (length(dim(new)) == length(dim(original))) # Nothing to do
+        return(orig.q.types)
+
+    # When one dimension remains, its dimname has been dropped.
+    # So need to infer questiontype.
+    if (length(dim(new)) == 1) {
+        # Try to match the names to the dimnames of the original array
+        matching.dims <- vapply(dimnames(original),
+                                FUN = function(x) all(names(new) %in% x),
+                                FUN.VALUE = logical(1))
+        q.type.per.dim <- getQuestionTypeForEachDimension(orig.q.types)
+        q.type.of.matched.dim <- unique(q.type.per.dim[matching.dims])
+        if (length(q.type.of.matched.dim) == 1) {
+            # Uniquely matched names to an original dimension so can
+            # use the corresponding question type.
+            q.dim <- questionDimension(q.type.of.matched.dim)
+            return(dropQuestionTypeDimensions(q.type.of.matched.dim, q.dim - 1))
+        }
+        # Can't match so just do the best you can based on the statistic
+        new.q.type <- "PickAny"
+        stat <- attr(new, "statistic")
+        if (!is.null(stat)) {
+            if (!grepl("%", stat)) { # Not obviously categorical
+                new.q.type <- "NumberMulti"
+            }
+        }
+        return(new.q.type)
+    }
+
+    # Identify question types of remaining dimensions and drop appropriately
+    orig.dims <- dim(nameDimensionAttributes(original))
```

Comment:
Safe to leave out `nameDimensionAttributes` since it's set on `R/subscript-tables.R#L18`.

Reply:
done